### PR TITLE
SCI-6158: add 1D Datatable plotmode

### PR DESCRIPTION
**Pull request by tschoonj**
In order to support rendering tables for both 1D and 2D datasets, it
appears to be necessary to define separate modes for each rank, which
is a bit unfortunate as it results in two entries in the plot mode list.

### CODE REVIEW

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java [hunk 1]
```diff
@@ -5,52 +5,55 @@
...
 			}
 			
 			@Override
 			public String getName() {
-				// TODO Auto-generated method stub
 				return null;
 			}
 			
 			@Override
 			public void setVisible(boolean isVisible) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java [hunk 2]
```diff
@@ -62,25 +65,21 @@ public void setUserObject(Object userObject) {
 			
 			@Override
 			public void setDataName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java [hunk 3]
```diff
@@ -5,52 +5,55 @@
...
 			}
 			
 			@Override
 			public void setVisible(boolean isVisible) {
-				// TODO Auto-generated method stub
 				
 			}
 			
 			@Override
 			public void setUserTrace(boolean isUserTrace) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java [hunk 4]
```diff
@@ -5,52 +5,55 @@
...
 		IDataset d = data[0];
 		
+		d.setShape(-1, 1);
 		
 		ITableDataTrace trace = new ITableDataTrace() {
 			
 			private Object userObject = null;
 			
 			@Override
 			public void setName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java [hunk 5]
```diff
@@ -108,7 +105,6 @@ public IDataset getData() {
 			
 			@Override
 			public void dispose() {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+				return null;
+			}
+			
+			@Override
+			public void setVisible(boolean isVisible) {
+				
+			}
+			
+			@Override
+			public void setUserTrace(boolean isUserTrace) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	public void displayData(IDataset[] data, ITrace[] update, IPlottingSystem<?> system, Object userObject) throws Exception {
+		IDataset d = data[0];
+		
+		
+		ITableDataTrace trace = new ITableDataTrace() {
+			
+			private Object userObject = null;
+			
+			@Override
+			public void setName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+				return null;
+			}
+			
+			@Override
+			public IDataset getData() {
+				return d;
+			}
+			
+			@Override
+			public void dispose() {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+				
+			}
+			
+			@Override
+			public String getName() {
+				return null;
+			}
+			
+			@Override
+			public void setVisible(boolean isVisible) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+			}
+			
+			@Override
+			public void setUserObject(Object userObject) {
+				this.userObject = userObject;
+				
+			}
+			
+			@Override
+			public void setDataName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/PlotModeDataTable2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

**Comment by gerring** [comment]
SonarQube analysis reported 10 issues
* ![CRITICAL][CRITICAL] 10 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

**Comment by jacobfilik** [comment]
"In order to support rendering tables for both 1D and 2D datasets, it
appears to be necessary to define separate modes for each rank"

Possibly not true. I have not looked at what happens if the number of labels (eg "X" or "Y" or "Row" or "Column") is larger than the minimum rank of the dataset, but we should probably just make it work for that case.

**Comment by tschoonj** [comment]
Not saying it's impossible, but it would be a lot of work to get it working with just one mode serving both 1D and 2D. I tried to set minimum rank to 1 in the 2D version and couldn't even select a single axis  for a 2D or 3D dataset in the Display column. Saw also some really ugly NPEs when using this for 1D datasets...

Would you agree to merging it in its current state for the upcoming release, while perhaps revisiting this for a later release? I really like the data tables... 😄

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		return null;
+	}
+
+	@Override
+	public IDataset getData() {
+		return d;
+	}
+
+	@Override
+	public void dispose() {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+		
+	}
+
+	@Override
+	public String getName() {
+		return null;
+	}
+
+	@Override
+	public void setVisible(boolean isVisible) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	@Override
+	public void setUserObject(Object userObject) {
+		this.userObject = userObject;
+		
+	}
+
+	@Override
+	public void setDataName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+		return null;
+	}
+
+	@Override
+	public void setVisible(boolean isVisible) {
+		
+	}
+
+	@Override
+	public void setUserTrace(boolean isUserTrace) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+public class TableDataTrace1D implements ITableDataTrace {
+	private final IDataset d;
+	private Object userObject = null;
+
+	public TableDataTrace1D(IDataset d) {
+		this.d = d;
+	}
+
+	@Override
+	public void setName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace1D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+		return null;
+	}
+
+	@Override
+	public IDataset getData() {
+		return d;
+	}
+
+	@Override
+	public void dispose() {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+		
+	}
+
+	@Override
+	public String getName() {
+		return null;
+	}
+
+	@Override
+	public void setVisible(boolean isVisible) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	@Override
+	public void setUserObject(Object userObject) {
+		this.userObject = userObject;
+		
+	}
+
+	@Override
+	public void setDataName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+		return null;
+	}
+
+	@Override
+	public void setVisible(boolean isVisible) {
+		
+	}
+
+	@Override
+	public void setUserTrace(boolean isUserTrace) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+public class TableDataTrace2D implements ITableDataTrace {
+	private final IDataset d;
+	private Object userObject = null;
+
+	public TableDataTrace2D(IDataset d) {
+		this.d = d;
+	}
+
+	@Override
+	public void setName(String name) {
```
Comment (gerring) [inline org.dawnsci.datavis.model/src/org/dawnsci/datavis/model/TableDataTrace2D.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

**Pull request merged**